### ISSUE TITLE
How does Redisson handle concurrent requests?

### ISSUE THREAD

**Issue by jonyt** [redisson/redisson#2675]
Say I have a page which requires two requests to render. In my session I hold a variable A=1. I use Redisson with `readMode="MEMORY" updateMode="AFTER_REQUEST" broadcastSessionEvents="false"`. 

Now I load the page. Both requests fire concurrently, the first completes before the second. The first request goes to Redis, gets A=1, changes it to A=2 and persists the session to Redis. The second request goes to Redis, gets A=1, does nothing and persists the session to Redis. Now in Redis I have A=1, even though I expected A=2. 

Does Redisson do anything clever to handle this scenario or is it up to me?

**Comment by mrniko** [comment]
This case is handled by developer

**Comment by mrniko** [comment]
> The second request goes to Redis, gets A=1, does nothing and persists the session to Redis. 

Redisson persists only changes.

**Comment by mrniko** [comment]
> A=1, changes it to A=2 and persists the session to Redis

First request delete A attribute from memory in all Redisson session instances. Thus further requests to this attribute will fetch it from Redis into memory again.

**Comment by msevcenko** [comment]
> First request delete A attribute from memory in all Redisson session instances. Thus further requests to this attribute will fetch it from Redis into memory again.

Looking at implementation, I'm not sure if this is what happens. When request terminates in AFTER_REQUEST, MEMORY mode, RedissonSession.save is called via UpdateValve. This method broadcasts AttributesPutAllMessage, which causes other instances to load new attributes.

Anyway, if the broadcast would work as you said, I don't think it would solve the issue. If the request not modifying the session, is terminated out of order, per AFTER_REQUEST+MEMORY policy, the shared redis session would get overwritten by outdated version of the session. Because there is no way how to detect that the request actually did not change the session. So broadcasting other instances to reload the (wrong version) of the session, wouldn't help, IMO, would it?

(I am a bit confused by the fact that RedissonSession extends StandardSession and delegates some operations to the base class. In particular, I don't see what actually is in StandardSession.attributes, which is loaded to RedissonSession.attrs, and later is stored to redis).

**Comment by msevcenko** [comment]
I once again checked the updated documentation:

> In `readMode=MEMORY` **all session attributes are always stored into Redis** after the end of the request regardless of the `Session.setAttribute` method invocation. It is useful in case when some objects stored in session change their own state without `Session.setAttribute` method execution. **Updated attributes are removed** from all other Session instances if `broadcastSessionUpdates=true` and reloaded from Redis when these attributes are requested.  

The first sentence is clear. **All** attributes must be saved, precisely because this policy is used in situation when the manager is unable to detect which attributes are _updated_.

So how the manager can notify about _updated_ attributes?

**Comment by msevcenko** [comment]
Please have a look at the following test, which demonstrates the discussed situation.

TL;DR, the test runs two requests in parallel, one which changes session attribute, the other that does not touch the session at all.

The result _may_ be, that the redis ends up with the wrong (old value). It is subject to race condition. Sometimes the request modifying the session may save the session later. Sometimes the request not modifying the session, may get the broadcast from the another, and update the session to the correct one before saving.

I do not see anything you can do with this issue implementation-wise, since you are not able to detect whether request did or did not change the session.

On application level, you may somehow filter requests that you know do not touch the session. But may you? Without specializing the RedissonSessionManager? I think the current implementation has no such feature.

And/or, you should provide a reliable sticky session mechanism, to reduce manipulation of the same session by different tomcat instances at the same time.

Anyway, I find this behavior rather dangerous and absolutely unexpected and undesirable, which is at least worth warning in the documentation, what do you think?

```java
package org.redisson.tomcat;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import org.apache.http.client.fluent.Executor;
import org.apache.http.client.fluent.Request;
import org.apache.http.impl.client.BasicCookieStore;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class RaceConditionTest {

    private void prepare(String contextName) throws IOException {
        String basePath = "src/test/webapp/META-INF/";
        Files.deleteIfExists(Paths.get(basePath + "context.xml"));
        Files.copy(Paths.get(basePath + contextName), Paths.get(basePath + "context.xml"));
    }
    
    @Test
    public void testUpdate_race() throws Exception {
        prepare("context_memory_after_request.xml");
        TomcatServer server1 = new TomcatServer("myapp", 8080, "src/test/");
        TomcatServer server2 = new TomcatServer("myapp", 8081, "src/test/");
        try {
            
            server1.start();
            server2.start();
 
            Executor executor = Executor.newInstance();
            BasicCookieStore cookieStore = new BasicCookieStore();
            executor.use(cookieStore);
            
            // create empty session, session is broadcast also to 8081
            start(8080, executor).call();
            // write session value, is broadcast also to 8081
            write(8080, executor, "test", "one").call();
            
            ExecutorService exec = Executors.newFixedThreadPool(4);
            
            // change the session to another value on one instance
            // do ping on another instance (request not modifying session)
            List<Future<String>> result = exec.invokeAll(Arrays.asList(
                    write(8080, executor, "test", "two"),
                    ping(8081, executor)
                    ));
            
            // wait for all requests
            Assertions.assertTrue(result.stream()
                    .map(t -> {
                        try {
                            return t.get();
                        } catch (InterruptedException | ExecutionException e) {
                            throw new RuntimeException(e);
                        }
                    })
                    .allMatch(r -> r.equals("OK")));
            
            // launch third server to see what is really in redis
            TomcatServer server3 = new TomcatServer("myapp", 8082, "src/test/");
            server3.start();
            String read = read(8082, executor, "test").call();
            // value two is expected
            Assertions.assertEquals("two", read);
            

        } finally {
            Executor.closeIdleConnections();
            server1.stop();
        }
    }

    private Callable<String> write(int port, Executor executor, String key, String value) {
        return () -> {
            String url = "http://localhost:" + port + "/myapp/write?key=" + key + "&value=" + value;
            return exec(executor, url);
        };
    }

    private Callable<String> ping(int port, Executor executor) throws IOException {
        return () -> {
            String url = "http://localhost:" + port + "/myapp/ping";
            return exec(executor, url);
        };
    }
    
    private Callable<String> start(int port, Executor executor) throws IOException {
        return () -> {
            String url = "http://localhost:" + port + "/myapp/start";
            return exec(executor, url);
        };
    }
    
    private Callable<String> read(int port, Executor executor, String key) throws IOException {
        return () -> {
            String url = "http://localhost:" + port + "/myapp/read?key=" + key;
            return exec(executor, url);
        };
    }

    private String exec(Executor executor, String url) {
        try {
            return executor.execute(Request.Get(url)).returnContent().asString();
        } catch (IOException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }
}
```
